common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -512,11 +510,82 @@ public void run() {
     }
 
     /**
-     * Sets up and shows dialog when user has exceeded the max attempts allowed to enter their YubiKey PIN.
+     * Attempts to authenticate using smartcard certificate.
+     * @param cert Chosen certificate for authentication.
+     * @param pin char array containing PIN.
+     * @param slotAlias Name of slot where certificate is located.
+     * @param piv A PivSession created from a SmartCardConnection that can interact with certificates located in the PIV slots on the YubiKey.
+     * @param request ClientCertRequest received from AzureActiveDirectoryWebViewClient.onReceivedClientCertRequest.
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private void useSmartcardCertForAuth(@NonNull final X509Certificate cert,
+                                         @NonNull final char[] pin,
+                                         @NonNull final String slotAlias,
+                                         @NonNull final PivSession piv,
+                                         @NonNull final ClientCertRequest request) {
+        final String methodTag = TAG + "useSmartcardCertForAuth:";
+        //Need to add a PivProvider instance to the beginning of the array of Security providers in order for signature logic to occur.
+        //Note that this provider is removed when the UsbYubiKeyDevice connection is closed.
```

Comment:
Very likely that we'll need to invoke [clearClientCertPreferences](https://developer.android.com/reference/android/webkit/WebView#clearClientCertPreferences(java.lang.Runnable)) when the connection is closed as well.

Reply:
Would you mind playing around with the 'result cached' scenarios?

Reply:
hmm I don't detect any issues when testing with logging in multiple accounts in a row with both types of CBA and a mix. Since `clearClientCertPreferences` is called every time `WebViewAuthorizationFragment` is created or a back button is pressed while in `WebViewAuthorizationFragment`, I think it should be able to handle both scenarios.